java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/config/domain/BuildContext.java

```diff
@@ -123,13 +133,53 @@ public void writeDockerFiles() throws IOException {
     try (BufferedWriter writer = Files.newBufferedWriter(dockerFilePath)) {
       writer.write(dockerfile.toString());
     }
+  }
+
+  private void writeDockerIgnore() throws IOException {
+    // If there's nothing to write, return
+    if (dockerignore.getLines().size() < 1) {
+      logger.debug("Skipping .dockerignore generation - nothing to write!");
+      return;
+    }
 
-    // write .dockerignore file, appending to an existing file if it exists
     Path dockerIgnorePath = workspaceDir.resolve(DOCKERIGNORE_NAME);
+    Set<String> existingDockerignoreLines = new HashSet<>();
+    boolean shouldPrependNewline = false;
+
+    // Read the contents of an existing .dockerignore file to avoid duplicating lines
+    if (Files.exists(dockerIgnorePath)) {
+      try (BufferedReader reader = Files.newBufferedReader(dockerIgnorePath)) {
+        reader.lines().forEach(existingDockerignoreLines::add);
+      }
+
+      // Check if the file ends in a UNIX newline character
+      RandomAccessFile file = new RandomAccessFile(dockerIgnorePath.toFile(), "r");
+      byte[] buffer = new byte[1];
+      file.seek(file.length() - 1);
+      file.read(buffer, 0, 1);
+      shouldPrependNewline = buffer[0] != '\n';
```

Comment:
👍  I actually think it might just be cleaner/less complex to always write a newline before our first addition to the dockerignore.